File: web/source/settings/lib/types/domain-permission.ts
/*
	GoToSocial
	Copyright (C) GoToSocial Authors admin@gotosocial.org
	SPDX-License-Identifier: AGPL-3.0-or-later

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU Affero General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU Affero General Public License for more details.

	You should have received a copy of the GNU Affero General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import typia from "typia";
import { PermType } from "./perm";
import { Links } from "parse-link-header";
import { PermSubContentType } from "./permsubcontenttype";

export const validateDomainPerms = typia.createValidate<DomainPerm[]>();

/**
 * A single domain permission entry (block, allow, draft, ignore).
 */
export interface DomainPerm {
	id?: string;
	domain: string;
	obfuscate?: boolean;
	private_comment?: string;
	public_comment?: string;
	created_at?: string;
	created_by?: string;
	subscription_id?: string;

	// Keys that should be stripped before
	// sending the domain permission (if imported).

	permission_type?: PermType;
	key?: string;
	suggest?: string;
	valid?: boolean;
	checked?: boolean;
	commentType?: string;
	replace_private_comment?: boolean;
	replace_public_comment?: boolean;
}

/**
 * Domain permissions mapped to an Object where the Object
 * keys are the "domain" value of each DomainPerm.
 */
export interface MappedDomainPerms {
	[key: string]: DomainPerm;
}

const domainPermStripOnImport: Set<keyof DomainPerm> = new Set([
	"key",
	"permission_type",
	"suggest",
	"valid",
	"checked",
	"commentType",
	"replace_private_comment",
	"replace_public_comment",
]);

/**
 * Returns true if provided DomainPerm Object key is one
 * that should be stripped when importing a domain permission.
 * 
 * @param key 
 * @returns 
 */
export function stripOnImport(key: keyof DomainPerm) {
	return domainPermStripOnImport.has(key);
}

export interface ImportDomainPermsParams {
	domains: DomainPerm[];

	// Internal processing keys;
	// remove before serdes of form.
	obfuscate?: boolean;
	commentType?: string;
	permType: PermType;
}

/**
 * Model domain permissions bulk export params.
 */
export interface ExportDomainPermsParams {
	permType: PermType;
	action: "export" | "export-file";
	exportType: "json" | "csv" | "plain";
}

/**
 * Parameters for GET to /api/v1/admin/domain_permission_drafts.
 */
export interface DomainPermDraftSearchParams {
	/**
	 * Show only drafts created by the given subscription ID.
	 */
	subscription_id?: string;
	/**
	 * Return only drafts that target the given domain.
	 */
	domain?: string;
	/**
	 * Filter on "block" or "allow" type drafts.
	 */
	permission_type?: PermType;
	/**
	 * Return only items *OLDER* than the given max ID (for paging downwards).
	 * The item with the specified ID will not be included in the response.
	 */
	max_id?: string;
	/**
	 * Return only items *NEWER* than the given since ID.
	 * The item with the specified ID will not be included in the response.
	 */
	since_id?: string;
	/**
	 * Return only items immediately *NEWER* than the given min ID (for paging upwards).
	 * The item with the specified ID will not be included in the response.
	 */
	min_id?: string;
	/**
	 * Number of items to return.
	 */
	limit?: number;
}

export interface DomainPermDraftSearchResp {
	drafts: DomainPerm[];
	links: Links | null;
}

export interface DomainPermDraftCreateParams {
	/**
	 * Domain to create the permission draft for.
	 */
	domain: string;
	/**
	 * Create a draft "allow" or a draft "block".
	 */
	permission_type: PermType;
	/**
	 * Obfuscate the name of the domain when serving it publicly.
	 * Eg., `example.org` becomes something like `ex***e.org`.
	 */
	obfuscate?: boolean;
	/**
	 * Public comment about this domain permission. This will be displayed
	 * alongside the domain permission if you choose to share permissions.
	 */
	public_comment?: string;
	/**
	 * Private comment about this domain permission.
	 * Will only be shown to other admins, so this is a useful way of
	 * internally keeping track of why a certain domain ended up permissioned.
	 */
	private_comment?: string;
}

/**
 * Parameters for GET to /api/v1/admin/domain_permission_excludes.
 */
export interface DomainPermExcludeSearchParams {
	/**
	 * Return only excludes that target the given domain.
	 */
	domain?: string;
	/**
	 * Return only items *OLDER* than the given max ID (for paging downwards).
	 * The item with the specified ID will not be included in the response.
	 */
	max_id?: string;
	/**
	 * Return only items *NEWER* than the given since ID.
	 * The item with the specified ID will not be included in the response.
	 */
	since_id?: string;
	/**
	 * Return only items immediately *NEWER* than the given min ID (for paging upwards).
	 * The item with the specified ID will not be included in the response.
	 */
	min_id?: string;
	/**
	 * Number of items to return.
	 */
	limit?: number;
}

export interface DomainPermExcludeSearchResp {
	excludes: DomainPerm[];
	links: Links | null;
}

export interface DomainPermExcludeCreateParams {
	/**
	 * Domain to create the permission exclude for.
	 */
	domain: string;
	/**
	 * Private comment about this domain permission.
	 * Will only be shown to other admins, so this is a useful way of
	 * internally keeping track of why a certain domain ended up permissioned.
	 */
	private_comment?: string;
}

/**
 * API model of one domain permission susbcription.
 */
export interface DomainPermSub {
	/**
	 * The ID of the domain permission subscription.
	 */
	id: string;
	/**
	 * The priority of the domain permission subscription.
	 */
	priority: number;
	/**
	 *  Time at which the subscription was created (ISO 8601 Datetime).
	 */
	created_at: string;
	/**
	 * Title of this subscription, as set by admin who created or updated it.
	 */
	title: string;
	/**
	 * The type of domain permission subscription (allow, block).
	 */
	permission_type: PermType;
	/**
	 * If true, domain permissions arising from this subscription will be created as drafts that must be approved by a moderator to take effect.
	 * If false, domain permissions from this subscription will come into force immediately.
	 */
	as_draft: boolean;
	/**
	 * If true, this domain permission subscription will "adopt" domain permissions
	 * which already exist on the instance, and which meet the following conditions:
	 * 1) they have no subscription ID (ie., they're "orphaned") and 2) they are present
	 * in the subscribed list. Such orphaned domain permissions will be given this
	 * subscription's subscription ID value and be managed by this subscription.
	 */
	adopt_orphans: boolean;
	/**
	 * ID of the account that created this subscription.
	 */
	created_by: string;
	/**
	 * URI to call in order to fetch the permissions list.
	 */
	uri: string;
	/**
	 * MIME content type to use when parsing the permissions list.
	 */
	content_type: PermSubContentType;
	/**
	 * (Optional) username to set for basic auth when doing a fetch of URI.
	 */
	fetch_username?: string;
	/**
	 * (Optional) password to set for basic auth when doing a fetch of URI.
	 */
	fetch_password?: string;
	/**
	 * Time at which the most recent fetch was attempted (ISO 8601 Datetime).
	 */
	fetched_at?: string;
	/**
	 *  Time of the most recent successful fetch (ISO 8601 Datetime).
	 */
	successfully_fetched_at?: string;
	/**
	 * If most recent fetch attempt failed, this field will contain an error message related to the fetch attempt.
	 */
	error?: string;
	/**
	 * Count of domain permission entries discovered at URI on last (successful) fetch.
	 */
	count: number;
}

/**
 * Parameters for GET to /api/v1/admin/domain_permission_subscriptions.
 */
export interface DomainPermSubSearchParams {
	/**
	 * Return only block or allow subscriptions.
	 */
	permission_type?: PermType;
	/**
	 * Return only items *OLDER* than the given max ID (for paging downwards).
	 * The item with the specified ID will not be included in the response.
	 */
	max_id?: string;
	/**
	 * Return only items *NEWER* than the given since ID.
	 * The item with the specified ID will not be included in the response.
	 */
	since_id?: string;
	/**
	 * Return only items immediately *NEWER* than the given min ID (for paging upwards).
	 * The item with the specified ID will not be included in the response.
	 */
	min_id?: string;
	/**
	 * Number of items to return.
	 */
	limit?: number;
}

export interface DomainPermSubCreateUpdateParams {
	/**
	 * The priority of the domain permission subscription.
	 */
	priority?: number;
	/**
	 * Title of this subscription, as set by admin who created or updated it.
	 */
	title?: string;
	/**
	 * URI to call in order to fetch the permissions list.
	 */
	uri: string;
	/**
	 * MIME content type to use when parsing the permissions list.
	 */
	content_type: PermSubContentType;
	/**
	 * If true, domain permissions arising from this subscription will be created as drafts that must be approved by a moderator to take effect.
	 * If false, domain permissions from this subscription will come into force immediately.
	 */
	as_draft?: boolean;
	/**
	 * If true, this domain permission subscription will "adopt" domain permissions
	 * which already exist on the instance, and which meet the following conditions:
	 * 1) they have no subscription ID (ie., they're "orphaned") and 2) they are present
	 * in the subscribed list. Such orphaned domain permissions will be given this
	 * subscription's subscription ID value and be managed by this subscription.
	 */
	adopt_orphans?: boolean;
	/**
	 * (Optional) username to set for basic auth when doing a fetch of URI.
	 */
	fetch_username?: string;
	/**
	 * (Optional) password to set for basic auth when doing a fetch of URI.
	 */
	fetch_password?: string;
	/**
	 * The type of domain permission subscription to create or update (allow, block).
	 */
	permission_type: PermType;
}

export interface DomainPermSubSearchResp {
	subs: DomainPermSub[];
	links: Links | null;
}

File: internal/api/client/admin/domainpermission.go
// GoToSocial
// Copyright (C) GoToSocial Authors admin@gotosocial.org
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package admin

import (
	"context"
	"errors"
	"fmt"
	"mime/multipart"
	"net/http"

	"github.com/gin-gonic/gin"
	apimodel "github.com/superseriousbusiness/gotosocial/internal/api/model"
	apiutil "github.com/superseriousbusiness/gotosocial/internal/api/util"
	"github.com/superseriousbusiness/gotosocial/internal/gtserror"
	"github.com/superseriousbusiness/gotosocial/internal/gtsmodel"
	"github.com/superseriousbusiness/gotosocial/internal/util"
)

type singleDomainPermCreate func(
	context.Context,
	gtsmodel.DomainPermissionType, // block/allow
	*gtsmodel.Account, // admin account
	string, // domain
	bool, // obfuscate
	string, // publicComment
	string, // privateComment
	string, // subscriptionID
) (*apimodel.DomainPermission, string, gtserror.WithCode)

type multiDomainPermCreate func(
	context.Context,
	gtsmodel.DomainPermissionType, // block/allow
	*gtsmodel.Account, // admin account
	*multipart.FileHeader, // domains
) (*apimodel.MultiStatus, gtserror.WithCode)

// createDomainPemissions either creates a single domain
// permission entry (block/allow) or imports multiple domain
// permission entries (multiple blocks, multiple allows)
// using the given functions.
//
// Handling the creation of both types of permissions in
// one function in this way reduces code duplication.
func (m *Module) createDomainPermissions(
	c *gin.Context,
	permType gtsmodel.DomainPermissionType,
	single singleDomainPermCreate,
	multi multiDomainPermCreate,
) {
	// Scope differs based on permType.
	var requireScope apiutil.Scope
	if permType == gtsmodel.DomainPermissionBlock {
		requireScope = apiutil.ScopeAdminWriteDomainBlocks
	} else {
		requireScope = apiutil.ScopeAdminWriteDomainAllows
	}

	authed, errWithCode := apiutil.TokenAuth(c,
		true, true, true, true,
		requireScope,
	)
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	if !*authed.User.Admin {
		err := fmt.Errorf("user %s not an admin", authed.User.ID)
		apiutil.ErrorHandler(c, gtserror.NewErrorForbidden(err, err.Error()), m.processor.InstanceGetV1)
		return
	}

	if authed.Account.IsMoving() {
		apiutil.ForbiddenAfterMove(c)
		return
	}

	if _, err := apiutil.NegotiateAccept(c, apiutil.JSONAcceptHeaders...); err != nil {
		apiutil.ErrorHandler(c, gtserror.NewErrorNotAcceptable(err, err.Error()), m.processor.InstanceGetV1)
		return
	}

	importing, errWithCode := apiutil.ParseDomainPermissionImport(c.Query(apiutil.DomainPermissionImportKey), false)
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	// Parse + validate form.
	form := new(apimodel.DomainPermissionRequest)
	if err := c.ShouldBind(form); err != nil {
		apiutil.ErrorHandler(c, gtserror.NewErrorBadRequest(err, err.Error()), m.processor.InstanceGetV1)
		return
	}

	var err error
	if importing && form.Domains.Size == 0 {
		err = errors.New("import was specified but list of domains is empty")
	} else if !importing && form.Domain == "" {
		err = errors.New("no domain provided")
	}

	if err != nil {
		apiutil.ErrorHandler(c, gtserror.NewErrorBadRequest(err, err.Error()), m.processor.InstanceGetV1)
		return
	}

	if !importing {
		// Single domain permission creation.
		perm, _, errWithCode := single(
			c.Request.Context(),
			permType,
			authed.Account,
			form.Domain,
			util.PtrOrZero(form.Obfuscate),
			util.PtrOrZero(form.PublicComment),
			util.PtrOrZero(form.PrivateComment),
			"", // No sub ID for single perm creation.
		)

		if errWithCode != nil {
			apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
			return
		}

		apiutil.JSON(c, http.StatusOK, perm)
		return
	}

	// We're importing multiple domain permissions,
	// so we're looking at a multi-status response.
	multiStatus, errWithCode := multi(
		c.Request.Context(),
		permType,
		authed.Account,
		form.Domains, // Pass the file through.
	)
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	// TODO: Return 207 and multiStatus data nicely
	//       when supported by the admin panel.
	if multiStatus.Metadata.Failure != 0 {
		failures := make(map[string]any, multiStatus.Metadata.Failure)
		for _, entry := range multiStatus.Data {
			failures[entry.Resource.(string)] = entry.Message
		}

		err := fmt.Errorf("one or more errors importing domain %ss: %+v", permType.String(), failures)
		apiutil.ErrorHandler(c, gtserror.NewErrorUnprocessableEntity(err, err.Error()), m.processor.InstanceGetV1)
		return
	}

	// Success, return slice of newly-created domain perms.
	domainPerms := make([]any, 0, multiStatus.Metadata.Success)
	for _, entry := range multiStatus.Data {
		domainPerms = append(domainPerms, entry.Resource)
	}

	apiutil.JSON(c, http.StatusOK, domainPerms)
}

func (m *Module) updateDomainPermission(
	c *gin.Context,
	permType gtsmodel.DomainPermissionType,
) {
	// Scope differs based on permType.
	var requireScope apiutil.Scope
	if permType == gtsmodel.DomainPermissionBlock {
		requireScope = apiutil.ScopeAdminWriteDomainBlocks
	} else {
		requireScope = apiutil.ScopeAdminWriteDomainAllows
	}

	authed, errWithCode := apiutil.TokenAuth(c,
		true, true, true, true,
		requireScope,
	)
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	if !*authed.User.Admin {
		err := fmt.Errorf("user %s not an admin", authed.User.ID)
		apiutil.ErrorHandler(c, gtserror.NewErrorForbidden(err, err.Error()), m.processor.InstanceGetV1)
		return
	}

	if authed.Account.IsMoving() {
		apiutil.ForbiddenAfterMove(c)
		return
	}

	if _, err := apiutil.NegotiateAccept(c, apiutil.JSONAcceptHeaders...); err != nil {
		apiutil.ErrorHandler(c, gtserror.NewErrorNotAcceptable(err, err.Error()), m.processor.InstanceGetV1)
		return
	}

	permID, errWithCode := apiutil.ParseID(c.Param(apiutil.IDKey))
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	// Parse + validate form.
	form := new(apimodel.DomainPermissionRequest)
	if err := c.ShouldBind(form); err != nil {
		apiutil.ErrorHandler(c, gtserror.NewErrorBadRequest(err, err.Error()), m.processor.InstanceGetV1)
		return
	}

	if form.Obfuscate == nil &&
		form.PrivateComment == nil &&
		form.PublicComment == nil {
		const errText = "empty form submitted"
		errWithCode := gtserror.NewErrorBadRequest(errors.New(errText), errText)
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	perm, errWithCode := m.processor.Admin().DomainPermissionUpdate(
		c.Request.Context(),
		permType,
		permID,
		form.Obfuscate,
		form.PublicComment,
		form.PrivateComment,
		nil, // Can't update perm sub ID this way yet.
	)
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	apiutil.JSON(c, http.StatusOK, perm)
}

// deleteDomainPermission deletes a single domain permission (block or allow).
func (m *Module) deleteDomainPermission(
	c *gin.Context,
	permType gtsmodel.DomainPermissionType, // block/allow
) {
	// Scope differs based on permType.
	var requireScope apiutil.Scope
	if permType == gtsmodel.DomainPermissionBlock {
		requireScope = apiutil.ScopeAdminWriteDomainBlocks
	} else {
		requireScope = apiutil.ScopeAdminWriteDomainAllows
	}

	authed, errWithCode := apiutil.TokenAuth(c,
		true, true, true, true,
		requireScope,
	)
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	if !*authed.User.Admin {
		err := fmt.Errorf("user %s not an admin", authed.User.ID)
		apiutil.ErrorHandler(c, gtserror.NewErrorForbidden(err, err.Error()), m.processor.InstanceGetV1)
		return
	}

	if authed.Account.IsMoving() {
		apiutil.ForbiddenAfterMove(c)
		return
	}

	if _, err := apiutil.NegotiateAccept(c, apiutil.JSONAcceptHeaders...); err != nil {
		apiutil.ErrorHandler(c, gtserror.NewErrorNotAcceptable(err, err.Error()), m.processor.InstanceGetV1)
		return
	}

	domainPermID, errWithCode := apiutil.ParseID(c.Param(apiutil.IDKey))
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	domainPerm, _, errWithCode := m.processor.Admin().DomainPermissionDelete(
		c.Request.Context(),
		permType,
		authed.Account,
		domainPermID,
	)
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	apiutil.JSON(c, http.StatusOK, domainPerm)
}

// getDomainPermission gets a single domain permission (block or allow).
func (m *Module) getDomainPermission(
	c *gin.Context,
	permType gtsmodel.DomainPermissionType,
) {
	// Scope differs based on permType.
	var requireScope apiutil.Scope
	if permType == gtsmodel.DomainPermissionBlock {
		requireScope = apiutil.ScopeAdminReadDomainBlocks
	} else {
		requireScope = apiutil.ScopeAdminReadDomainAllows
	}

	authed, errWithCode := apiutil.TokenAuth(c,
		true, true, true, true,
		requireScope,
	)
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	if !*authed.User.Admin {
		err := fmt.Errorf("user %s not an admin", authed.User.ID)
		apiutil.ErrorHandler(c, gtserror.NewErrorForbidden(err, err.Error()), m.processor.InstanceGetV1)
		return
	}

	if _, err := apiutil.NegotiateAccept(c, apiutil.JSONAcceptHeaders...); err != nil {
		apiutil.ErrorHandler(c, gtserror.NewErrorNotAcceptable(err, err.Error()), m.processor.InstanceGetV1)
		return
	}

	domainPermID, errWithCode := apiutil.ParseID(c.Param(apiutil.IDKey))
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	export, errWithCode := apiutil.ParseDomainPermissionExport(c.Query(apiutil.DomainPermissionExportKey), false)
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	domainPerm, errWithCode := m.processor.Admin().DomainPermissionGet(
		c.Request.Context(),
		permType,
		domainPermID,
		export,
	)
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	apiutil.JSON(c, http.StatusOK, domainPerm)
}

// getDomainPermissions gets all domain permissions of the given type (block, allow).
func (m *Module) getDomainPermissions(
	c *gin.Context,
	permType gtsmodel.DomainPermissionType,
) {
	// Scope differs based on permType.
	var requireScope apiutil.Scope
	if permType == gtsmodel.DomainPermissionBlock {
		requireScope = apiutil.ScopeAdminReadDomainBlocks
	} else {
		requireScope = apiutil.ScopeAdminReadDomainAllows
	}

	authed, errWithCode := apiutil.TokenAuth(c,
		true, true, true, true,
		requireScope,
	)
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	if !*authed.User.Admin {
		err := fmt.Errorf("user %s not an admin", authed.User.ID)
		apiutil.ErrorHandler(c, gtserror.NewErrorForbidden(err, err.Error()), m.processor.InstanceGetV1)
		return
	}

	if _, err := apiutil.NegotiateAccept(c, apiutil.JSONAcceptHeaders...); err != nil {
		apiutil.ErrorHandler(c, gtserror.NewErrorNotAcceptable(err, err.Error()), m.processor.InstanceGetV1)
		return
	}

	export, errWithCode := apiutil.ParseDomainPermissionExport(c.Query(apiutil.DomainPermissionExportKey), false)
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	domainPerm, errWithCode := m.processor.Admin().DomainPermissionsGet(
		c.Request.Context(),
		permType,
		authed.Account,
		export,
	)
	if errWithCode != nil {
		apiutil.ErrorHandler(c, errWithCode, m.processor.InstanceGetV1)
		return
	}

	apiutil.JSON(c, http.StatusOK, domainPerm)
}

// parseDomainPermissionType is a util function to parse i
// to a DomainPermissionType, or return a suitable error.
func parseDomainPermissionType(i string) (
	permType gtsmodel.DomainPermissionType,
	errWithCode gtserror.WithCode,
) {
	if i == "" {
		const errText = "permission_type not set, must be one of block or allow"
		errWithCode = gtserror.NewErrorBadRequest(errors.New(errText), errText)
		return
	}

	permType = gtsmodel.ParseDomainPermissionType(i)
	if permType == gtsmodel.DomainPermissionUnknown {
		var errText = fmt.Sprintf("permission_type %s not recognized, must be one of block or allow", i)
		errWithCode = gtserror.NewErrorBadRequest(errors.New(errText), errText)
	}

	return
}

// parseDomainPermSubContentType is a util function to parse i
// to a DomainPermSubContentType, or return a suitable error.
func parseDomainPermSubContentType(i string) (
	contentType gtsmodel.DomainPermSubContentType,
	errWithCode gtserror.WithCode,
) {
	if i == "" {
		const errText = "content_type not set, must be one of text/csv, text/plain or application/json"
		errWithCode = gtserror.NewErrorBadRequest(errors.New(errText), errText)
		return
	}

	contentType = gtsmodel.NewDomainPermSubContentType(i)
	if contentType == gtsmodel.DomainPermSubContentTypeUnknown {
		var errText = fmt.Sprintf("content_type %s not recognized, must be one of text/csv, text/plain or application/json", i)
		errWithCode = gtserror.NewErrorBadRequest(errors.New(errText), errText)
	}

	return
}
